src/gpio.rs

```diff
@@ -293,1040 +377,841 @@ macro_rules! gpio {
-                            exti.ftsr.modify(|r, w| unsafe { w.bits(r.bits() | (1 << self.i)) });
-                        }
-                    }
-                }
-
-                /// Enable external interrupts from this pin.
-                fn enable_interrupt(&mut self, exti: &EXTI) {
-                    exti.imr.modify(|r, w| unsafe { w.bits(r.bits() | (1 << self.i)) });
-                }
-
-                /// Disable external interrupts from this pin
-                fn disable_interrupt(&mut self, exti: &EXTI) {
-                    exti.imr.modify(|r, w| unsafe { w.bits(r.bits() & !(1 << self.i)) });
-                }
-
-                /// Clear the interrupt pending bit for this pin
-                fn clear_interrupt_pending_bit(&mut self) {
-                    unsafe { (*EXTI::ptr()).pr.write(|w| w.bits(1 << self.i) ) };
-                }
-
-                /// Reads the interrupt pending bit for this pin
-                fn check_interrupt(&mut self) -> bool {
-                    unsafe { ((*EXTI::ptr()).pr.read().bits() & (1 << self.i)) != 0 }
-                }
-            }
-
-            impl <MODE> StatefulOutputPin for Generic<Output<MODE>> {
-                fn is_set_high(&self) -> Result<bool, Self::Error> {
-                    self.is_set_low().map(|b| !b)
-                }
-
-                fn is_set_low(&self) -> Result<bool, Self::Error> {
-                    // NOTE(unsafe) atomic read with no side effects
-                    Ok(unsafe { (*$GPIOX::ptr()).odr.read().bits() & (1 << self.i) == 0 })
+            impl<MODE> PEPin<MODE, $port_id> {
+                pub fn erase(self) -> EPin<MODE> {
+                    EPin::$PXx(self)
                 }
             }
 
-            impl <MODE> toggleable::Default for Generic<Output<MODE>> {}
-
-            impl InputPin for Generic<Output<OpenDrain>> {
-                type Error = Infallible;
-                fn is_high(&self) -> Result<bool, Self::Error> {
-                    self.is_low().map(|b| !b)
-                }
-
-                fn is_low(&self) -> Result<bool, Self::Error> {
-                    // NOTE(unsafe) atomic read with no side effects
-                    Ok(unsafe { (*$GPIOX::ptr()).idr.read().bits() & (1 << self.i) == 0 })
+            impl<MODE, CR, const N: u8> Pin<MODE, CR, $port_id, N>
+            where
+                MODE: Active,
+            {
+                /// Erases the pin number and port from the type
+                ///
+                /// This is useful when you want to collect the pins into an array where you
+                /// need all the elements to have the same type
+                pub fn erase(self) -> EPin<MODE> {
+                    self.erase_number().erase()
                 }
             }
+        }
+    }
+}
 
-            pub type $PXx<MODE> = Pxx<MODE>;
-
-
-            impl<MODE> Mode<MODE> for Generic<MODE> {}
+/// Pin
+pub struct Pin<MODE, CR, const P: char, const N: u8> {
```

Comment:
Another naming question: Should we stick with P and N here, or just write out `Port` and `Number`. The latter is clearly more verbose, but also makes it less ambiguous

Reply:
It will be too long to use it everywhere. I will add it to docs better.